File: PythonDesignPatterns/adopter.py
"""

                    +------------------+
                    |   Application    |
                    +------------------+
                    | + play_song()    |
                    +------------------+
                             |
                             | uses
                             v

                    +------------------+
                    |    <<Target>>    |
                    |      Player      |
                    +------------------+
                    | + play(file)     |
                    +------------------+
                       ^            ^
                       |            |
             implements|            |implements
                       |            |

        +--------------------+   +--------------------+
        | Mp3PlayerAdapter   |   | WavPlayerAdapter   |
        +--------------------+   +--------------------+
        | - mp3_player       |   | - wav_player       |
        +--------------------+   +--------------------+
        | + play(file)       |   | + play(file)       |
        +--------------------+   +--------------------+
                 |                         |
                 | adapts                  | adapts
                 v                         v

        +--------------------+   +--------------------+
        |     Mp3Player      |   |     WavPlayer      |
        +--------------------+   +--------------------+
        | + play_mp3(file)   |   | + play_wav(file)   |
        +--------------------+   +--------------------+

"""

print(__doc__)


class Mp3Player:
    def play_mp3(self, filename):
        print(f"Playing MP3: {filename}")


class WavPlayer:
    def play_wav(self, filename):
        print(f"Playing WAV: {filename}")


# Client
class Application:
    def play_song(self, player, filename):
        player.play(filename)


# Adapter
class Mp3PlayerAdapter:
    def __init__(self, mp3_player):
        self.mp3_player = mp3_player

    def play(self, filename):
        self.mp3_player.play_mp3(filename)


# Adapter
class WavPlayerAdapter:
    def __init__(self, wav_player):
        self.wav_player = wav_player

    def play(self, filename):
        self.wav_player.play_wav(filename)


# Existing objects (Adaptees)
mp3_player = Mp3Player()
wav_player = WavPlayer()

# Adapters
mp3_adapter = Mp3PlayerAdapter(mp3_player)
wav_adapter = WavPlayerAdapter(wav_player)

# Client code
app = Application()

app.play_song(mp3_adapter, "robo")
app.play_song(wav_adapter, "east or west")

File: NodeDesignPatterns/adopter.ts

let doc=`

                    +------------------+
                    |   Application    |
                    +------------------+
                    | + play_song()    |
                    +------------------+
                             |
                             | uses
                             v

                    +------------------+
                    |    <<Target>>    |
                    |      Player      |
                    +------------------+
                    | + play(file)     |
                    +------------------+
                       ^            ^
                       |            |
             implements|            |implements
                       |            |

        +--------------------+   +--------------------+
        | Mp3PlayerAdapter   |   | WavPlayerAdapter   |
        +--------------------+   +--------------------+
        | - mp3_player       |   | - wav_player       |
        +--------------------+   +--------------------+
        | + play(file)       |   | + play(file)       |
        +--------------------+   +--------------------+
                 |                         |
                 | adapts                  | adapts
                 v                         v

        +--------------------+   +--------------------+
        |     Mp3Player      |   |     WavPlayer      |
        +--------------------+   +--------------------+
        | + play_mp3(file)   |   | + play_wav(file)   |
        +--------------------+   +--------------------+


`
console.log(doc)


class Mp3Player {
    playMp3(filename: string): void {
        console.log(`Playing MP3: ${filename}`);
    }
}

class WavPlayer {
    playWav(filename: string): void {
        console.log(`Playing WAV: ${filename}`);
    }
}

interface Player {
    play(filename: string): void;
}

class Mp3PlayerAdapter implements Player {
    constructor(private mp3Player: Mp3Player) {}

    play(filename: string): void {
        this.mp3Player.playMp3(filename);
    }
}

class WavPlayerAdapter implements Player {
    constructor(private wavPlayer: WavPlayer) {}

    play(filename: string): void {
        this.wavPlayer.playWav(filename);
    }
}


class Application {
    playSong(player: Player, filename: string): void {
        player.play(filename);
    }
}


const mp3Player = new Mp3Player();
const wavPlayer = new WavPlayer();

const mp3Adapter = new Mp3PlayerAdapter(mp3Player);
const wavAdapter = new WavPlayerAdapter(wavPlayer);

const app = new Application();

app.playSong(mp3Adapter, "robo");
app.playSong(wavAdapter, "east or west");